### DISCUSSION TITLE
How do I implement Quasi-Fermi approximation for drift-diffusion problem in fipy?

### DISCUSSION THREAD

**Discussion by sbtristan98** [usnistgov/fipy#1070] [Q&A]
Hello, 

I am trying to solve a drift-diffusion problem for a solar cell in fipy.
These are the currently implemented equations:

`
#Continuity Equation For Electrons
eq1 = (self.gen_rate - (1.00 / -q) * (ExponentialConvectionTerm(coeff=q * self.nmob * phi.grad, var=n) + DiffusionTerm(coeff=q * self.nmob * D, var=n)))

#Poisson Equation
eq2 = (DiffusionTerm(coeff=epsilon, var=phi) + (- n) * q / (epsilon_0) == 0)

eq = eq1 & eq2
`

To implement the QF approximation we need this equation: "n = Nc * numerix.exp((χ - Φn + phi)/D)" or like this in logarithmic form "Φn = χ + phi - D * numerix.log((n.value / Nc))".

As one can see, in the current two equations we solved for n and phi, however in the QF approximation, I should solve for the state variables Φn and phi.

However, I can't find any appropriate way to define this change of variables since I'm not aware of a way of setting an exponential inside a convection term. Is there a solution for this?

Cheers,

Tristan

**Comment by guyer** [comment]
There isn't an implicit way to set the exponential inside the convection term. You need to run the derivatives through by hand:

$$
\begin{aligned}
\nabla\cdot\left(q \mu_n D \nabla n\right) &= \nabla\cdot\left\\{q \mu_n D \nabla \left[N_C \exp\frac{\xi - \phi_n + \phi}{D}\right]\right\\}
\\
&= \nabla\cdot\left\\{q \mu_n D N_C \exp\frac{\xi - \phi_n + \phi}{D} \nabla \frac{\xi - \phi_n + \phi}{D}\right\\}
\\
&= \nabla\cdot\left\\{q \mu_n D n \nabla \frac{\xi - \phi_n + \phi}{D}\right\\}
\\
&= \nabla\cdot\left\\{q \mu_n n \nabla \left(\xi - \phi_n + \phi\right)\right\\}
\end{aligned}
$$

If you're careful with signs, combining this with the first term should leave you a single DiffusionTerm in $\phi_n$, but I haven't looked at this stuff in a long time and the transformations between hole/electron concentrations, pseudo-Fermi levels, and Slotboom variables always made my head hurt.
